File: src/cpu/activation.ts
export interface CPUActivationFn {
  activate(val: number): number;
  prime(val: number): number;
}

/**
 * Linear activation function f(x) = x
 */
export class Linear implements CPUActivationFn {
  activate(val: number): number {
    return val;
  }

  prime(_val: number): number {
    return 1;
  }
}

/**
 * Sigmoid activation function f(x) = 1 / (1 + e^(-x))
 */
export class Sigmoid implements CPUActivationFn {
  activate(val: number): number {
    return 1 / (1 + Math.exp(-val));
  }

  prime(val: number): number {
    return val * (1 - val);
  }
}

/**
 * Tanh activation function f(x) = (e^x - e^-x) / (e^x + e^-x)
 * This is the same as the sigmoid function, but is more robust to outliers
 */
export class Tanh implements CPUActivationFn {
  activate(val: number): number {
    return Math.tanh(val);
  }

  prime(val: number): number {
    return 1 - (val * val);
  }
}

/**
 * ReLU activation function f(x) = max(0, x)
 * This is a rectified linear unit, which is a smooth approximation to the sigmoid function.
 */
export class Relu implements CPUActivationFn {
  activate(val: number): number {
    return Math.max(0, val);
  }

  prime(val: number): number {
    return val > 0 ? 1 : 0;
  }
}

/**
 * Relu6 activation function f(x) = min(max(0, x), 6)
 * This is a rectified linear unit with a 6-value output range.
 */
export class Relu6 implements CPUActivationFn {
  activate(val: number): number {
    return Math.min(Math.max(0, val), 6);
  }

  prime(val: number): number {
    return val > 0 ? 1 : 0;
  }
}

/**
 * Leaky ReLU activation function f(x) = x if x > 0, 0.01 * x otherwise
 */
export class LeakyRelu implements CPUActivationFn {
  activate(val: number): number {
    return val > 0 ? val : 0.01 * val;
  }

  prime(val: number): number {
    return val > 0 ? 1 : 0.01;
  }
}

/**
 * Elu activation function f(x) = x if x >= 0, 1.01 * (e^x - 1) otherwise
 * This is a rectified linear unit with an exponential output range.
 */
export class Elu implements CPUActivationFn {
  activate(val: number): number {
    return val >= 0 ? val : Math.exp(val) - 1;
  }

  prime(val: number): number {
    return val > 0 ? 1 : Math.exp(val);
  }
}

/**
 * Selu activation function f(x) = x if x >= 0, 1.67 * (e^x - 1) otherwise
 * This is a scaled version of the Elu function, which is a smoother approximation to the ReLU function.
 */
export class Selu implements CPUActivationFn {
  activate(val: number): number {
    return val >= 0 ? val : 1.0507 * (Math.exp(val) - 1);
  }

  prime(val: number): number {
    return val > 0 ? 1 : 1.0507 * Math.exp(val);
  }
}

File: rust/src/cpu/activation.rs
use crate::Activation;
pub struct CPUActivation {
    pub activation: Activation,
    pub activate: fn(x: &f32) -> f32,
    pub prime: fn(x: &f32) -> f32,
}

impl CPUActivation {
    pub fn from(activation: Activation) -> CPUActivation {
        match activation {
            Activation::Sigmoid => CPUActivation {
                activation: Activation::Sigmoid,
                activate: sigmoid,
                prime: sigmoid_prime,
            },
            Activation::Tanh => CPUActivation {
                activation: Activation::Tanh,
                activate: tanh,
                prime: tanh_prime,
            },
            Activation::Linear => CPUActivation {
                activation: Activation::Linear,
                activate: linear,
                prime: linear_prime,
            },
            Activation::Relu => CPUActivation {
                activation: Activation::Relu,
                activate: relu,
                prime: relu_prime,
            },
            Activation::Relu6 => CPUActivation {
                activation: Activation::Relu6,
                activate: relu6,
                prime: relu6_prime,
            },
            Activation::LeakyRelu => CPUActivation {
                activation: Activation::LeakyRelu,
                activate: leaky_relu,
                prime: leaky_relu_prime,
            },
            Activation::Elu => CPUActivation {
                activation: Activation::Elu,
                activate: elu,
                prime: elu_prime,
            },
            Activation::Selu => CPUActivation {
                activation: Activation::Selu,
                activate: selu,
                prime: selu_prime,
            },
        }
    }

    pub fn from_option(activation: Option<Activation>) -> Option<CPUActivation> {
        if let Some(activation) = activation {
            Some(CPUActivation::from(activation))
        } else {
            None
        }
    }

    pub fn memoize_output(activation: &CPUActivation) -> bool {
        match activation.activation {
            Activation::Sigmoid | Activation::Tanh => true,
            _ => true
        }
    }
}

fn sigmoid(x: &f32) -> f32 {
    return 1.0 / (1.0 + (-x).exp());
}

fn sigmoid_prime(x: &f32) -> f32 {
    return x * (1.0 - x);
}

fn tanh(x: &f32) -> f32 {
    return x.tanh();
}

fn tanh_prime(x: &f32) -> f32 {
    return 1.0 - tanh(x).powi(2);
}

fn linear(x: &f32) -> f32 {
    return *x;
}

fn linear_prime(_x: &f32) -> f32 {
    return 1.0;
}

fn relu(x: &f32) -> f32 {
    return x.max(0.0);
}

fn relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.0 };
}

fn relu6(x: &f32) -> f32 {
    return x.max(0.0).min(6.0);
}

fn relu6_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.0 };
}

fn leaky_relu(x: &f32) -> f32 {
    return if *x > 0.0 { *x } else { x.max(0.01 * x) };
}

fn leaky_relu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 0.01 };
}

fn elu(x: &f32) -> f32 {
    return if *x >= 0.0 { *x } else { x.exp() - 1.0 };
}

fn elu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { x.exp() };
}

fn selu(x: &f32) -> f32 {
    return if *x >= 0.0 { *x } else { 1.0507 * (x.exp() - 1.0) };
}

fn selu_prime(x: &f32) -> f32 {
    return if *x > 0.0 { 1.0 } else { 1.0507 * x.exp() };
}